Add State pattern, HFSM and pushdown automaton demo

NathanLovato opened the pull request:
Currently in 2d folder as it's a 2d project. There's no folder that really fits atm.

closes #250

bojidar-bg commented on 2d/finite_state_machine/debug/keys.gd:
```diff
@@ -0,0 +1,6 @@
+
+
+func _ready():
+	set_as_toplevel(true)
```
This file is the same as explanations.gd, differing only in what they extend.
Consider making just one script which extends `Control` (where `set_as_toplevel` is defined).

bojidar-bg commented:
@NathanLovato Well, you should get notified by `git` whenever you try to add such a file.

NathanLovato commented:
I simplified the code: it uses signals instead of static delegates. The character has more states as well now, showing attacks that combine with the jump.

NathanLovato commented:
Reworked and simplified the code once more. It's much cleaner now, with better separation of concerns. Still works exactly the same as before. Could someone check it and approve for merge if you find it good enough?

akien-mga commented:
I've only had a very cursory look but it seems good to me :)
